src/omv/ports/stm32/jpeg.c

```diff
@@ -0,0 +1,753 @@
+    HAL_JPEG_Decode_DMA(&JPEG_Handle,
+                        (uint8_t *) JPEG_in_data, IM_MIN(JPEG_in_data_len, MAX_BLOCK_SIZE),
+                        mcu_row_buffer, IM_MIN(dst_w_mcus_bytes, MAX_BLOCK_SIZE));
+
+    for (int y_offset = 0; y_offset < src->h; y_offset += mcu_h) {
+        int h = y_offset / mcu_h;
+        uint8_t *this_mcu_row_buffer_ptr = mcu_row_buffer + (dst_w_mcus_bytes * (h % 2));
+        uint8_t *next_mcu_row_buffer_ptr = mcu_row_buffer + (dst_w_mcus_bytes * ((h + 1) % 2));
+
+        int dy = src->h - y_offset;
+        if (dy > mcu_h) {
+            dy = mcu_h;
+        }
+
+        if ((y_offset + mcu_h) < src->h) {
+            // not last row
+            // Invalidate the MCU row for DMA.
+            SCB_InvalidateDCache_by_Addr((uint32_t *) next_mcu_row_buffer_ptr, dst_w_mcus_bytes);
+        }
+
+        bool abort = false;
+
+        // Wait for the MCUs to be processed.
+        for (mp_uint_t tick_start = mp_hal_ticks_ms(); !JPEG_output_paused; ) {
+            __WFI();
+
+            // If we haven't exited this loop before the timeout then we need to abort the transfer.
+            if ((mp_hal_ticks_ms() - tick_start) > DECODE_TIMEOUT) {
+                HAL_JPEG_Abort(&JPEG_Handle);
+                abort = true;
+                break;
+            }
+        }
+
+        if (abort) {
+            break;
+        }
+
+        if ((y_offset + mcu_h) < src->h) {
+            // not last row
+            // Reset the lock.
+            JPEG_output_paused = false;
+            // Restart the DMA process on the next row of MCUs.
+            HAL_JPEG_ConfigOutputBuffer(&JPEG_Handle,
+                                        next_mcu_row_buffer_ptr, IM_MIN(dst_w_mcus_bytes, MAX_BLOCK_SIZE));
+            HAL_JPEG_Resume(&JPEG_Handle, JPEG_PAUSE_RESUME_OUTPUT);
+        }
+
+        if (JPEG_Info.ColorSpace == JPEG_GRAYSCALE_COLORSPACE) {
+            for (int x_offset = 0; x_offset < src->w; x_offset += MCU_W) {
+                uint8_t *Y0 = this_mcu_row_buffer_ptr + (x_offset * MCU_H);
+
+                int dx = src->w - x_offset;
+                if (dx > MCU_W) {
+                    dx = MCU_W;
+                }
+
+                switch (dst->pixfmt) {
+                    case PIXFORMAT_BINARY: {
+                        for (int y = y_offset, yy = y + dy; y < yy; y++) {
+                            uint32_t *rp = IMAGE_COMPUTE_BINARY_PIXEL_ROW_PTR(dst, y);
+
+                            for (int x = x_offset, xx = x + dx; x < xx; x++) {
+                                int p = *Y0++;
+                                int v = p > 128;
+                                IMAGE_PUT_BINARY_PIXEL_FAST(rp, x, v);
+                            }
+
+                            Y0 += MCU_W - dx;
+                        }
+                        break;
+                    }
+                    case PIXFORMAT_GRAYSCALE: {
+                        for (int y = y_offset, yy = y + dy; y < yy; y++) {
```

Comment:
Please no more of this.
```suggestion
                        for (int y = y_offset; y < dy + y_offset; y++) {
```